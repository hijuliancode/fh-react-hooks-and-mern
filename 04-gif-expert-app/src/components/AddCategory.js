import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({ setCategories }) => {
  const inputSearchElm = document.getElementById('searchCategory')
  const [ inputSearchValue, setInputSearchValue ] = useState('');

  const handleInputChange = (e) => {
    setInputSearchValue( e.target.value )
  }

  const handleSubmit = (e) => {
    e.preventDefault();

    if ( inputSearchValue.trim().length > 2 ) {
      let newCategory = {
        id: guidGenerator(),
        name: inputSearchValue
      }
      setCategories( cats => [newCategory, ...cats] )

      setInputSearchValue('')
      inputSearchElm.focus()
    }
  }

  function guidGenerator() { // Globally Unique IDentifier
    var S4 = function() {
       return (((1+Math.random())*0x10000)|0).toString(16).substring(1);
    };
    return (S4()+S4()+"-"+S4()+"-"+S4()+"-"+S4()+"-"+S4()+S4()+S4());
  }

  return (
    <form onSubmit={ handleSubmit }>
      <h2 className="mb-1 is-size-6">Add Category</h2>
      <div className="field has-addons mb-4 input-search">
        <div className="control">
          <input
            className="input"
            id="searchCategory"
            type="text"
            placeholder="Find a repository"
            value={ inputSearchValue }
            onChange={ handleInputChange } />
        </div>
        <div className="control">
          <button className="button is-info" type="submit">
            Search
          </button>
        </div>
      </div>
    </form>
  )
}

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
}