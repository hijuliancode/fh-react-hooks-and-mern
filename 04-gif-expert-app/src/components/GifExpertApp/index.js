import React, { useState } from 'react';

const GifExpertAppComponent = () => {
  const [ categories, setCategories ] = useState([
    { id: 'OnePunch', name: 'One Punch'},
    { id: 'SamuraiX', name: 'Samurai X'},
    { id: 'DragonBall', name: 'Dragon Ball'},
  ])

  const handleAdd = () => {
    let inputSearch = document.getElementById('searchCategory')
    let inputSearchValue = inputSearch.value

    if ( inputSearchValue ) {
      let newCategorie = {
        id: guidGenerator(),
        name: inputSearchValue
      }
      setCategories( cats => [...cats, newCategorie] )

      // Reset Input
      inputSearch.value = ''
      inputSearch.focus()
    }
  }

  function guidGenerator() { // Globally Unique IDentifier
    var S4 = function() {
       return (((1+Math.random())*0x10000)|0).toString(16).substring(1);
    };
    return (S4()+S4()+"-"+S4()+"-"+S4()+"-"+S4()+"-"+S4()+S4()+S4());
}

  return (
    <div className="container">
      <h2 className="is-size-5-mobile is-size-4 has-text-centered mt-4">Gif Expert App</h2>
      <hr />

      <div className="field has-addons mb-4 input-search">
        <div className="control">
          <input className="input" id="searchCategory" type="text" placeholder="Find a repository" />
        </div>
        <div className="control">
          <button onClick={() => handleAdd()} className="button is-info" type="button">
            Search
          </button>
        </div>
      </div>

      {
        categories.map( (category) => {
          return <button key={category.id} className="button is-small is-primary is-outlined mr-1 mb-1">{category.name}</button>
        } )
      }

    </div>
  );
}
 
export default GifExpertAppComponent;