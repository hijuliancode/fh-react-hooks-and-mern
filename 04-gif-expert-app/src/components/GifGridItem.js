import React from 'react'
import PropTypes from "prop-types";

export const GifGridItem = ( { url, title, externalUrl } ) => {
  return (
    <div className="card animate__animated animate__fadeIn">
      <div className="card-image">
        <a href={ externalUrl } target="_blank" className="image is-4by3 animate__animated animate__fadeIn animate__fast" rel="noreferrer">
          <img src={ url } alt={ title } />
        </a>
      </div>
      <div className="card-content">
        <div className="media">
          <div className="media-content">
            <p className="title is-4">{ title }</p>
          </div>
        </div>
      </div>
    </div>
  )
}

GifGridItem.propTypes = {
  url: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  externalUrl: PropTypes.string.isRequired,
}