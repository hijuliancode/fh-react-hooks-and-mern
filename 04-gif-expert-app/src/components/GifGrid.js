import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem'

export const GifGrid = ({ category }) => {

  const { data:images, loading } = useFetchGifs(category)

  return (
    <>
      { // If is loading
        loading && (
          <div class="notification animate__delay-1s is-success animate__animated animate__flash animate__infinite">
            <p>Cargando...</p>
          </div>
        )
      }
      <h2 className="title is-5 is-capitalized">{ category }</h2>
      <div className="columns is-multiline is-tablet mb-6">
        {
          images.map( image => (
            <div key={ image.id } className="column is-one-quarter">
              <GifGridItem { ...image } />
            </div>
          ))
        }
      </div>
    </>
  )
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
}