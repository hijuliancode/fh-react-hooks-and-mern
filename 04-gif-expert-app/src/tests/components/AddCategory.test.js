import React from 'react'
import { shallow } from 'enzyme'
import { AddCategory }  from '../../components/AddCategory'

describe('Test in AddCategory.js', () => {
  const setCategories = () => {}
  test('should render <AddCategory /> correctly', () => {
    const wrapper = shallow( <AddCategory setCategories={ setCategories } /> )
    expect( wrapper ).toMatchSnapshot()
  })
})
