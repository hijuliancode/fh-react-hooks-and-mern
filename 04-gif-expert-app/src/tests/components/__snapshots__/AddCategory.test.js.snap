// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in AddCategory.js should render <AddCategory /> correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <h2
    className="mb-1 is-size-6"
  >
    Add Category
  </h2>
  <div
    className="field has-addons mb-4 input-search"
  >
    <div
      className="control"
    >
      <input
        className="input"
        id="searchCategory"
        onChange={[Function]}
        placeholder="Find a repository"
        type="text"
        value=""
      />
    </div>
    <div
      className="control"
    >
      <button
        className="button is-info"
        type="submit"
      >
        Search
      </button>
    </div>
  </div>
</form>
`;
